src/util/nft.js

```diff
@@ -283,6 +285,68 @@ export const getNFTClassesRespectDualPrefix = async (axios, owner) => {
   return arraysOfNFTClasses.flat();
 };
 
+async function getPurchasePrice(axios, nftId, classId, txHash) {
+  try {
+    const { data } = await axios.get(api.getChainRawTx(txHash));
+    const { price } = data.tx.body.messages.find(
+      m =>
+        m['@type'] === '/likechain.likenft.v1.MsgBuyNFT' &&
+        m.class_id === classId &&
+        m.nft_id === nftId
+    );
+    return new BigNumber(price).shiftedBy(-9).toNumber();
+  } catch (error) {
+    // eslint-disable-next-line no-console
+    console.error(error);
+    return null;
+  }
+}
+
+export async function populatePurchasePriceFromChain(axios, history) {
+  const historyWithPrice = await Promise.all(
+    history.map(async e => {
+      const { event, txHash, classId, nftId } = e;
+      if (event === 'buy_nft') {
+        e.price = await getPurchasePrice(axios, nftId, classId, txHash);
+      }
+      return e;
+    })
+  );
+  return historyWithPrice;
+}
+
+export async function populateCollectPriceFromDB({
+  axios,
+  history,
+  classId,
+  nftId,
+}) {
+  try {
+    const eventMap = new Map();
+    history.forEach(e => {
+      eventMap.set(`${e.txHash}-${e.nftId}-${e.event}`, e);
+    });
+
+    const { data } = await axios.get(
+      api.getNFTHistory({
+        classId,
+        nftId,
+      })
+    );
+
+    const historyInDB = data.list;
+    historyInDB.forEach(e => {
+      const key = `${e.txHash}-${e.nftId}-${e.event}`;
+      if (eventMap.has(key)) {
+        eventMap.get(key).price = e.price;
+      }
+    });
```

Comment:
it seems this function does not return/output anything?

Reply:
It doesn't return because it intends to mutate the `history`  passed into the function.

Reply:
it did not modify history either, or did i miss anything?

Reply:
BTW, both `populatePurchasePriceFromChain` and `populateCollectPriceFromDB` mutate `history` here.

Reply:
BTW, both `populatePurchasePriceFromChain` and `populateCollectPriceFromDB` mutate `event` of `history` here.

Reply:
`e` from `eventMap` is also the item in `history`, so `eventMap.get(key).price = e.price;` mutates `history` here.
(probably I made it not easy to read at first glance 🌚

Reply:
i think we should explicitly clone history the return the new modified instance to make this easier to read

Reply:
Can this issue be fixed by better function naming? Since we want to `await` two async functions to mutate a single source of `history` here. If the functions return a new modified instance, the process would be sequential but not concurrent.